core/templates/dev/head/expressions/expressionSyntaxTree.js

```diff
@@ -0,0 +1,427 @@
+      ExprWrongNumArgsError.prototype.constructor = ExprWrongNumArgsError;
+      ExprWrongNumArgsError.prototype.name = 'ExprWrongNumArgsError';
+      ExprWrongNumArgsError.prototype.toString = function() {
+        return this.name + ': {' + this.args + '} not in range [' +
+          this.expectedMin + ',' + this.expectedMax + ']';
+      };
+
+      var ExprWrongArgTypeError = function(arg, actualType, expectedType) {
+        this.arg = arg;
+        this.actualType = actualType;
+        this.expectedType = expectedType;
+      };
+      ExprWrongArgTypeError.prototype = new ExpressionError();
+      ExprWrongArgTypeError.prototype.constructor = ExprWrongArgTypeError;
+      ExprWrongArgTypeError.prototype.name = 'ExprWrongArgTypeError';
+      ExprWrongArgTypeError.prototype.toString = function() {
+        return this.name + ': ' + this.arg + ' has type ' + this.actualType +
+          ' which does not match expected type ' + this.expectedType;
+      };
+
+      var getParamsUsedInExpression = function(expression) {
+        var _findParams = function(parseTree) {
+          var paramsFound = [];
+
+          if (parseTree instanceof Array) {
+            if (parseTree[0] === '#') {
+              paramsFound.push(parseTree[1]);
+            } else {
+              for (var i = 1; i < parseTree.length; i++) {
+                paramsFound = paramsFound.concat(_findParams(parseTree[i]));
+              }
+            }
+          }
+
+          var uniqueParams = [];
+          for (var i = 0; i < paramsFound.length; i++) {
+            if (uniqueParams.indexOf(paramsFound[i]) === -1) {
+              uniqueParams.push(paramsFound[i]);
+            }
+          }
+
+          return uniqueParams.sort();
+        };
+
+        var parsed = expressionParserService.parse(expression);
+        return _findParams(parsed);
+      };
+
+      // Checks if the args array has the expectedNum number of elements an
+      // throws an error if not. If optional expectedMax is specified, it
+      // verifies the number of args is in [expectedNum, expectedMax] range
+      // inclusive.
+      var verifyNumArgs = function(args, expectedNum, expectedMax) {
+        if (expectedMax === undefined) {
+          expectedMax = expectedNum;
+        }
+        if (args.length >= expectedNum && args.length <= expectedMax) {
+          return;
+        }
+        throw new ExprWrongNumArgsError(args, expectedNum, expectedMax);
+      };
+
+      var _verifyArgsHaveType = function(args, expectedType) {
+        for (var i = 0; i < args.length; i++) {
+          if (args[i] != expectedType) {
+            throw new ExprWrongArgTypeError(
+              originalValue, typeof originalValue, expectedType);
+          }
+        }
+        return true;
+      };
+
+      var _verifyArgsHaveSameType = function(arg1, arg2) {
+        if (arg1 != arg2) {
+          throw new ExprWrongArgTypeError(arg1, arg1, arg2);
```

Comment:
The second arg *is* the type of the argument. Notice how the function is trying to check that the "type" is the same between the two things set in, but it is actually just checking whether or not they are equal, because what is being sent into this function is a string that represents the evaluated type of the sub-expression in that spot. So the actual "type" that we're concerned about here is the same as the argument being sent into this function.